File: qt/src/core/VideoExtractor.cc
#include "VideoExtractor.hh"

#include <QCryptographicHash>
#include <QDir>
#include <QFileInfo>
#include <QStandardPaths>

#include "utils/Constants.hh"
#include "utils/PathUtils.hh"

namespace timelapse {

VideoExtractor::VideoExtractor(QObject* parent)
   : QObject(parent) {}

VideoExtractor::~VideoExtractor() {
   this->cancelExtraction();
}

auto VideoExtractor::extractFrames(QString const& videoPath, int32_t fps)
   -> std::expected<QString, Error> {
   if (this->_extracting) {
      return std::unexpected(Error{ErrorCode::ExtractionInProgress,
                                   "Extraction already in progress"});
   }

   // Find ffmpeg
   auto ffmpegPath = this->findFfmpeg();
   if (!ffmpegPath) {
      return std::unexpected(ffmpegPath.error());
   }

   // Get video filename for cache folder
   QFileInfo videoInfo(videoPath);
   QString videoFilename = videoInfo.fileName();

   // Check if already cached
   if (this->hasCachedFrames(videoFilename)) {
      return this->generateCacheFolderName(videoFilename);
   }

   // Create cache folder
   QString cacheFolderName = this->generateCacheFolderName(videoFilename);
   QString cacheFolderPath = PathUtils::cacheDir() + "/" + cacheFolderName;

   if (!PathUtils::ensureDir(cacheFolderPath)) {
      return std::unexpected(Error{ErrorCode::CacheCreationFailed,
                                   "Failed to create cache folder: " + cacheFolderPath});
   }

   // Build ffmpeg command
   QString outputPattern = cacheFolderPath + "/%05d.jpg";
   QStringList args;
   args << "-i" << videoPath
        << "-vf" << QString("fps=%1").arg(fps)
        << "-q:v" << "2"  // High quality JPEG
        << outputPattern;

   // Start ffmpeg process
   this->_ffmpegProcess = new QProcess(this);
   this->_extracting = true;

   emit extractionStarted(videoPath);

   this->_ffmpegProcess->start(*ffmpegPath, args);

   if (!this->_ffmpegProcess->waitForStarted(5000)) {
      this->_extracting = false;
      QString error = this->_ffmpegProcess->errorString();
      delete this->_ffmpegProcess;
      this->_ffmpegProcess = nullptr;
      return std::unexpected(Error{ErrorCode::FfmpegStartFailed,
                                   "Failed to start ffmpeg: " + error});
   }

   // Wait for completion (blocking for simplicity)
   // In a real app, you'd use signals for async handling
   if (!this->_ffmpegProcess->waitForFinished(-1)) {  // No timeout
      this->_extracting = false;
      QString error = this->_ffmpegProcess->errorString();
      delete this->_ffmpegProcess;
      this->_ffmpegProcess = nullptr;
      return std::unexpected(Error{ErrorCode::FfmpegFailed, "ffmpeg failed: " + error});
   }

   int exitCode = this->_ffmpegProcess->exitCode();
   this->_extracting = false;
   delete this->_ffmpegProcess;
   this->_ffmpegProcess = nullptr;

   if (exitCode != 0) {
      return std::unexpected(Error{ErrorCode::FfmpegExitError,
                                   "ffmpeg exited with code: " + QString::number(exitCode)});
   }

   emit extractionFinished(cacheFolderName);
   return cacheFolderName;
}

auto VideoExtractor::hasCachedFrames(QString const& videoFilename) const -> bool {
   QString cacheFolderPath = this->getCacheFolderPath(videoFilename);
   QDir cacheDir(cacheFolderPath);

   if (!cacheDir.exists()) {
      return false;
   }

   // Check if there are any jpg files
   QStringList filters;
   filters << "*.jpg";
   return !cacheDir.entryList(filters, QDir::Files).isEmpty();
}

auto VideoExtractor::getCacheFolderPath(QString const& videoFilename) const -> QString {
   return PathUtils::cacheDir() + "/" + this->generateCacheFolderName(videoFilename);
}

auto VideoExtractor::isExtracting() const -> bool {
   return this->_extracting;
}

void VideoExtractor::cancelExtraction() {
   if (this->_ffmpegProcess && this->_ffmpegProcess->state() != QProcess::NotRunning) {
      this->_ffmpegProcess->kill();
      this->_ffmpegProcess->waitForFinished(3000);
   }
   this->_extracting = false;
}

auto VideoExtractor::findFfmpeg() const -> std::expected<QString, Error> {
   // Check common locations
   QStringList searchPaths;

#ifdef Q_OS_MACOS
   searchPaths << "/opt/homebrew/bin/ffmpeg"
               << "/usr/local/bin/ffmpeg";
#endif

#ifdef Q_OS_LINUX
   searchPaths << "/usr/bin/ffmpeg"
               << "/usr/local/bin/ffmpeg";
#endif

#ifdef Q_OS_WIN
   searchPaths << "C:/ffmpeg/bin/ffmpeg.exe"
               << "ffmpeg.exe";  // Rely on PATH
#endif

   // Also search in PATH
   QString pathFfmpeg = QStandardPaths::findExecutable("ffmpeg");
   if (!pathFfmpeg.isEmpty()) {
      return pathFfmpeg;
   }

   for (auto const& path : searchPaths) {
      if (QFileInfo::exists(path)) {
         return path;
      }
   }

   return std::unexpected(Error{ErrorCode::FfmpegNotFound,
                                "ffmpeg not found. Please install ffmpeg."});
}

auto VideoExtractor::generateCacheFolderName(QString const& videoFilename) const
   -> QString {
   // Create a hash-based folder name to avoid special characters
   QByteArray hash = QCryptographicHash::hash(
      videoFilename.toUtf8(),
      QCryptographicHash::Md5);

   // Use first 8 chars of hex hash + sanitized filename
   QString safeName = videoFilename;
   safeName.replace(QRegularExpression("[^a-zA-Z0-9._-]"), "_");

   return hash.toHex().left(8) + "_" + safeName;
}

}  // namespace timelapse

File: rust/src/core/video_extractor.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use crate::core::error::{Error, ErrorCode, Result};
use crate::utils::constants::DEFAULT_EXTRACTION_FPS;
use crate::utils::paths;

pub struct VideoExtractor {
   extracting: bool,
}

impl VideoExtractor {
   pub fn new() -> Self {
      Self { extracting: false }
   }

   /// Extract frames from a video file
   pub fn extract_frames(&mut self, video_path: &Path, fps: Option<u32>) -> Result<String> {
      if self.extracting {
         return Err(Error::new(
            ErrorCode::ExtractionInProgress,
            "Extraction already in progress",
         ));
      }

      let fps = fps.unwrap_or(DEFAULT_EXTRACTION_FPS);

      // Find ffmpeg
      let ffmpeg_path = self.find_ffmpeg()?;

      // Get video filename for cache folder
      let video_filename = video_path
         .file_name()
         .and_then(|n| n.to_str())
         .unwrap_or("unknown");

      // Check if already cached
      if self.has_cached_frames(video_filename) {
         return Ok(self.generate_cache_folder_name(video_filename));
      }

      // Create cache folder
      let cache_folder_name = self.generate_cache_folder_name(video_filename);
      let cache_folder_path = paths::cache_dir().join(&cache_folder_name);

      if !paths::ensure_dir(&cache_folder_path) {
         return Err(Error::new(
            ErrorCode::CacheCreationFailed,
            format!("Failed to create cache folder: {:?}", cache_folder_path),
         ));
      }

      // Build output pattern
      let output_pattern = cache_folder_path.join("%05d.jpg");

      self.extracting = true;

      // Run ffmpeg
      let result = Command::new(&ffmpeg_path)
         .args([
            "-i",
            video_path.to_str().unwrap_or(""),
            "-vf",
            &format!("fps={}", fps),
            "-q:v",
            "2",
            output_pattern.to_str().unwrap_or(""),
         ])
         .output();

      self.extracting = false;

      match result {
         Ok(output) => {
            if output.status.success() {
               Ok(cache_folder_name)
            } else {
               let stderr = String::from_utf8_lossy(&output.stderr);
               Err(Error::new(
                  ErrorCode::FfmpegExitError,
                  format!(
                     "ffmpeg exited with code {:?}: {}",
                     output.status.code(),
                     stderr
                  ),
               ))
            }
         }
         Err(e) => Err(Error::new(
            ErrorCode::FfmpegStartFailed,
            format!("Failed to run ffmpeg: {}", e),
         )),
      }
   }

   /// Check if frames are already cached for a video
   pub fn has_cached_frames(&self, video_filename: &str) -> bool {
      let cache_path = self.get_cache_folder_path(video_filename);
      if !cache_path.exists() {
         return false;
      }

      // Check for any jpg files
      fs::read_dir(&cache_path)
         .map(|entries| {
            entries.filter_map(|e| e.ok()).any(|e| {
               e.path()
                  .extension()
                  .map(|ext| ext == "jpg")
                  .unwrap_or(false)
            })
         })
         .unwrap_or(false)
   }

   /// Get the cache folder path for a video
   pub fn get_cache_folder_path(&self, video_filename: &str) -> PathBuf {
      paths::cache_dir().join(self.generate_cache_folder_name(video_filename))
   }

   /// Check if extraction is in progress
   pub fn is_extracting(&self) -> bool {
      self.extracting
   }

   /// Find ffmpeg executable
   fn find_ffmpeg(&self) -> Result<PathBuf> {
      // Check common locations
      let search_paths = [
         "/opt/homebrew/bin/ffmpeg",
         "/usr/local/bin/ffmpeg",
         "/usr/bin/ffmpeg",
      ];

      for path in search_paths {
         let p = PathBuf::from(path);
         if p.exists() {
            return Ok(p);
         }
      }

      // Try to find in PATH
      if let Ok(output) = Command::new("which").arg("ffmpeg").output() {
         if output.status.success() {
            let path = String::from_utf8_lossy(&output.stdout).trim().to_string();
            if !path.is_empty() {
               return Ok(PathBuf::from(path));
            }
         }
      }

      Err(Error::new(
         ErrorCode::FfmpegNotFound,
         "ffmpeg not found. Please install ffmpeg.",
      ))
   }

   /// Generate a cache folder name from the video filename
   fn generate_cache_folder_name(&self, video_filename: &str) -> String {
      use std::collections::hash_map::DefaultHasher;
      use std::hash::{Hash, Hasher};

      let mut hasher = DefaultHasher::new();
      video_filename.hash(&mut hasher);
      let hash = hasher.finish();

      // Sanitize filename
      let safe_name: String = video_filename
         .chars()
         .map(|c| {
            if c.is_alphanumeric() || c == '.' || c == '_' || c == '-' {
               c
            } else {
               '_'
            }
         })
         .collect();

      format!("{:08x}_{}", hash as u32, safe_name)
   }
}

impl Default for VideoExtractor {
   fn default() -> Self {
      Self::new()
   }
}
